Moved all templates that use Release or CrateDetails to Tera

Kixiron opened the pull request:
This is unfortunately a really large PR and I apologize, but there wasn't really any other choice since they're all interdependent on how `Release` and `CrateDetails` serialize. I didn't delete any handlebars templates since at this stage all the inter-dependence is really fuzzy and hard to discern, so I think it's just not worth the risk of deleting an entire segment of the site. It should only be 2 or three more PRs after this, so it shouldn't be a huge deal to keep them around for a bit
This is best taken by commit, each commit deals with one single template (which can amount to multiple pages in some cases).

## [Commit 1](https://github.com/rust-lang/docs.rs/commit/96acd9ad74f9d9bf71ca4639c0f84f2a82027f6a)

<details>

## Release Feed

### Before

*Formatted for readability*

```xml
<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
    <title>Docs.rs</title>
    <subtitle>Recent Rust crates</subtitle>

    <link href="https://docs.rs/releases/feed" rel="self" />
    <link href="https://docs.rs/" />
    <link href="https://pubsubhubbub.appspot.com" rel="hub" />
    <link href="https://pubsubhubbub.superfeedr.com" rel="hub" />

    <id>urn:docs-rs:0.6.0-8b5fb70-2020-06-22</id>
    <updated>2020-06-22T21:37:04.673103+00:00</updated>

    <entry>
        <title>dyn-fmt-0.2.0</title>
        <link href="https://docs.rs/dyn-fmt/0.2.0/dyn_fmt/" />
        <id>urn:docs-rs:dyn-fmt:0.2.0</id>
        <updated>2020-06-22T21:37:04.673103+00:00</updated>
        <summary>Provides dynamic string format.</summary>
        <author>
            <name>docs.rs</name>
        </author>
    </entry>

</feed>
```

### After

*Pre-formatted*

```xml
<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
    <title>Docs.rs</title>
    <subtitle>Recent Rust crates</subtitle>

    <link href="https://docs.rs/releases/feed" rel="self" />
    <link href="https://docs.rs/" />
    <link href="https://pubsubhubbub.appspot.com" rel="hub" />
    <link href="https://pubsubhubbub.superfeedr.com" rel="hub" />

    <id>urn:docs-rs:0.6.0 (ceeb7aa 2020-06-22)</id>
    <updated>2020-05-10T04:01:29.304849+00:00</updated>

        <entry>
            <title>serde-1.0.109</title>

            <link href="&#x2F;crate&#x2F;serde&#x2F;1.0.109" />
            <id>urn:docs-rs:serde:1.0.109</id>
            <updated>2020-05-10T04:01:29.304849+00:00</updated>

            <summary>
                A generic serialization&#x2F;deserialization framework
            </summary>

            <author>
                <name>docs.rs</name>
            </author>
        </entry>
</feed>
```

</details>

## [Commit 2](https://github.com/rust-lang/docs.rs/commit/1073de82bc54f9d57e81a1d8648a478c28d8f151)

<details>

## Home

### Before

![image](https://user-images.githubusercontent.com/25047011/85338276-945ecb80-b4a7-11ea-97ae-1ce9b8f4474b.png)

### After

![image](https://user-images.githubusercontent.com/25047011/85338310-a80a3200-b4a7-11ea-8d2d-8ccf624ff33f.png)

</details>

## [Commit 3](https://github.com/rust-lang/docs.rs/commit/f1219ec5a78de38c6290a99606eea2c206de2fb7)

<details>

## Recent Releases

### Before

![image](https://user-images.githubusercontent.com/25047011/85338502-020af780-b4a8-11ea-869e-06c44d5b4bb3.png)

### After

![image](https://user-images.githubusercontent.com/25047011/85338480-fa4b5300-b4a7-11ea-96c0-d7192b7104e0.png)

## Releases by Stars

### Before

![image](https://user-images.githubusercontent.com/25047011/85338792-95dcc380-b4a8-11ea-9106-9a3e8f098883.png)

### After

![image](https://user-images.githubusercontent.com/25047011/85338806-9b3a0e00-b4a8-11ea-8adf-508c38c817fc.png)

## Recent Failures

### Before

![image](https://user-images.githubusercontent.com/25047011/85338836-a725d000-b4a8-11ea-9b4d-22aae4e40cbd.png)

### After

*My machine can't build crates for some reason*

## Failures by Stars

### Before

![image](https://user-images.githubusercontent.com/25047011/85338884-bc026380-b4a8-11ea-88b2-b377a536353d.png)

### After

*My machine can't build crates for some reason*

</details>

## [Commit 4](https://github.com/rust-lang/docs.rs/commit/a00abfb504049a1e2199ed1a221d14989b7adc2b)

<details>

## Authors

### Before

![image](https://user-images.githubusercontent.com/25047011/85338918-c58bcb80-b4a8-11ea-8979-bf9e1ab46e86.png)

### After

![image](https://user-images.githubusercontent.com/25047011/85338942-ce7c9d00-b4a8-11ea-8d7d-f193ead215f0.png)

</details>

## [Commit 5](https://github.com/rust-lang/docs.rs/commit/a284ece83586f7efe4978256f25bebf50ac86a1e)

<details>

## Search

### Before

![image](https://user-images.githubusercontent.com/25047011/85338993-dfc5a980-b4a8-11ea-82f5-f22f90f3d39f.png)

### After

![image](https://user-images.githubusercontent.com/25047011/85339009-e6ecb780-b4a8-11ea-8454-a5b178ab3d94.png)

</details>

## [Commit 6](https://github.com/rust-lang/docs.rs/commit/cc474aae269939af2621986343de55e6d2acb2ba)

<details>

## Failed Search

### Before

![image](https://user-images.githubusercontent.com/25047011/85339033-f2d87980-b4a8-11ea-95e4-c992895ad397.png)

### After

![image](https://user-images.githubusercontent.com/25047011/85339048-f966f100-b4a8-11ea-8a7e-28588bdd8531.png)

## Failed empty Search

### Before

![image](https://user-images.githubusercontent.com/25047011/85339072-01bf2c00-b4a9-11ea-982e-b97938e0ebdc.png)

### After

![image](https://user-images.githubusercontent.com/25047011/85339089-0a176700-b4a9-11ea-9e52-e830bdafdfdd.png)

</details>

jyn514 commented on src/web/releases.rs:
```diff
@@ -463,50 +450,54 @@ pub fn releases_failures_by_stars_handler(req: &mut Request) -> IronResult<Respo
-        packages.len() == RELEASES_IN_RELEASES as usize,
+        releases.len() == RELEASES_IN_RELEASES as usize,
         page_number != 1,
     );
```
Why did this comment go away? Do you know why it was there originally?

Kixiron replied:
Same as the other occurrence

jyn514 commented on src/web/page/web_page.rs:
```diff
@@ -3,6 +3,7 @@ use iron::{headers::ContentType, response::Response, status::Status, IronResult,
 use tera::Context;
 
+/// When making using a custom status, use a closure that coerces to a `fn(&Self) -> Status`
 #[macro_export]
```
```suggestion
```
We're not using this from any other crate, so there's no need to export it.

Kixiron replied:
It's needed because of macro scoping rules

Nemo157 replied:
It shouldn't be, you should `#[macro_use]` from above to get macros into context. `#[macro_export]` is _only_ for publicly exporting macros (and as a side-effect makes them visible from the root of the crate).

jyn514 commented:
> I didn't delete any handlebars templates since at this stage all the inter-dependence is really fuzzy and hard to discern, so I think it's just not worth the risk of deleting an entire segment of the site. It should only be 2 or three more PRs after this, so it shouldn't be a huge deal to keep them around for a bit

Oops, missed this the first time. It should be pretty clear I think when part of the site is missing since it will give 404s, if we don't catch it in the test suite that's a good opportunity to improve the test suite (we can test by hand since each template only has a few corresponding pages). So I think we should try to remove them as we go.

Kixiron commented on tera-templates/core/home.html:
```diff
@@ -0,0 +1,108 @@
+        <div class="recent-releases-container">
+            <div class="release">
+                <a href="/releases">
+                    <strong>Recent Releases</strong>
```
The content isn't actually that much the same, the entire template was essentially an `if home_page { .. } else { .. }`

Kixiron commented on src/web/releases.rs:
```diff
@@ -362,114 +362,102 @@ pub fn releases_feed_handler(req: &mut Request) -> IronResult<Response> {
+
+    let (description, number_releases, release_order) = match release_type {
+        ReleaseType::Recent => (
+            "Recently uploaded crates",
+            RELEASES_IN_RELEASES,
+            Order::ReleaseTime,
+        ),
+        ReleaseType::Stars => (
+            "Crates with most stars",
+            RELEASES_IN_RELEASES,
+            Order::GithubStars,
+        ),
+        ReleaseType::RecentFailures => (
+            "Recent crates failed to build",
+            RELEASES_IN_RELEASES,
+            Order::RecentFailures,
+        ),
+        ReleaseType::Failures => (
+            "Crates with most stars failed to build",
+            RELEASES_IN_RELEASES,
+            Order::FailuresByGithubStars,
+        ),
+
+        ReleaseType::Author | ReleaseType::Search => unreachable!(
+            "The authors and search page have special requirements and cannot use this handler",
+        ),
+    };
+
+    let releases = {
+        let conn = extension!(req, Pool).get()?;
+        get_releases(&conn, page_number, number_releases, release_order)
+    };
 
     // Show next and previous page buttons
-    // This is a temporary solution to avoid expensive COUNT(*)
     let (show_next_page, show_previous_page) = (
-        packages.len() == RELEASES_IN_RELEASES as usize,
+        releases.len() == RELEASES_IN_RELEASES as usize,
         page_number != 1,
     );
```
The comment said it was a temporary solution, but it works perfectly fine

Kixiron commented on tera-templates/releases/header.html:
```diff
@@ -0,0 +1,77 @@
+                            </a>
+                        </li>
+
+                        {%- if author -%}
+                            <li class="pure-menu-item">
+                                <a href="#" class="pure-menu-link{% if tab == 'author' %} pure-menu-active{% endif %}">
+                                    <i class="fa fa-fw fa-user"></i>
+                                    <span class="title"> {{ author }}</span>
+                                </a>
+                            </li>
+                        {%- endif -%}
```
I've noticed that a lot of things just break if there's no entries, like all the recent release pages explicitly errored if there were no releases

Kixiron commented:
The problem is that the current templates have almost no segmentation, everything at this stage is super intertwined. I can still try though, I just doubt it'll be fun

Kixiron commented on src/web/crate_details.rs:
```diff
@@ -17,7 +17,7 @@ use serde_json::Value;
 // TODO: Add target name and versions
 
-#[derive(Debug)]
+#[derive(Debug, Clone, PartialEq)]
```
They're just basic derives that had no reason to not be implemented

Kixiron commented on src/web/crate_details.rs:
```diff
@@ -106,7 +107,7 @@ impl Serialize for CrateDetails {
 }
 
-#[derive(Debug, Eq, PartialEq, Serialize)]
+#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
```
It's just a basic derive that had no reason to not be implemented

Kixiron commented on tera-templates/releases/feed.xml:
```diff
@@ -0,0 +1,39 @@
+            </summary>
+
+            <author>
+                <name>docs.rs</name>
```
I have no earthly idea, if you want it changed then now's the time

jyn514 replied:
Let's leave it alone for now, this feels like a separate project.

Kixiron commented on src/web/crate_details.rs:
```diff
@@ -42,6 +42,7 @@ pub struct CrateDetails {
     github_stars: Option<i32>,
     github_forks: Option<i32>,
     github_issues: Option<i32>,
+    // TODO: Is this even needed for rendering pages?
     pub(crate) metadata: MetaData,
```
I think what my comment may have been getting at is the `pub target_name: String,` field, which is (maybe?) `default_target` since the `pub(crate) doc_targets: Vec<String>,` field also exists

Nemo157 replied:
IIRC `target_name` is the name of the library (which can be different from the crate).

Kixiron commented on tera-templates/releases/header.html:
```diff
@@ -0,0 +1,77 @@
+{% macro header(title, description, tab, author=false) %}
+    <div class="cratesfyi-package-container">
+        <div class="container">
+            <h1 id="crate-title">{{ title }}</h1>
```
`header.hbs` was incorporated into `base.html`. Because of how handlebars works, that file acted as a literal copy+paste into the file. `header.hbs` contained all the meta, js and css that's now in `base.html` and then pasted in `navigation.hbs`, which was basically an if statement between "release mode" and "package mode" with those all being pulled into independent templates because they are indeed independent

Kixiron commented on tera-templates/core/home.html:
```diff
@@ -0,0 +1,108 @@
+                                    </div>
+                                {%- else -%}
+                                    <div class="pure-u-1 pure-u-sm-4-24 pure-u-md-3-24 date">
+                                        {{ release.release_time | timeformat(relative=true) }}
```
![image](https://user-images.githubusercontent.com/25047011/85960570-0f255c00-b96a-11ea-98bb-861b52102814.png)

Nemo157 commented on tera-templates/core/home.html:
```diff
@@ -0,0 +1,108 @@
+                                    </div>
+                                {%- else -%}
+                                    <div class="pure-u-1 pure-u-sm-4-24 pure-u-md-3-24 date">
+                                        {{ release.release_time | timeformat(relative=true) }}
```
Can we stick a `title` with the absolute time on everywhere there is a relative time?

Kixiron commented:
I addressed all comments and I changed the authors page to have stars again
![image](https://user-images.githubusercontent.com/25047011/86040781-3cc1e200-ba0a-11ea-9ab4-6e90cf0d40a4.png)
I also added a title to the authors page that shows the relative date to when the crate was published just for fun
![image](https://user-images.githubusercontent.com/25047011/86040860-5b27dd80-ba0a-11ea-9257-0c0e87a2f4e7.png)
The failure for the docker build came from #864, I fixed it in my final commit
```
COPY failed: stat /var/lib/docker/tmp/docker-builder032416435/docker/entrypoint.sh: no such file or directory
```

jyn514 commented:
> The failure for the docker build came from #864, I fixed it in my final commit (the docker folder was renamed to dockerfiles but this wasn't reflected in dockerfiles/Dockerfile)

Can you split that out into a separate PR so we can merge it right away?

Kixiron commented:
I opened #867 for an independent fix

jyn514 commented on src/web/releases.rs:
```diff
@@ -1040,11 +1042,128 @@ mod tests {
+            let web = env.frontend();
+            assert_success("/releases/queue", web)?;
+
+            env.db().fake_release().name("some_random_crate").create()?;
+            env.db()
+                .fake_release()
+                .name("some_random_crate_that_failed")
+                .build_result_successful(false)
+                .create()?;
+            assert_success("/releases/queue", web)
```
This doesn't test what you think it does. `fake_release` does not touch the build queue at all, this is the same test as if you hadn't had the call to `fake_release()`.

Kixiron replied:
Fixed to test against a populated queue in 3fff667

pietroalbini replied:
Hmm, I'd remove that test -- I added a better one in #863, and your copy uses an API that my PR removed.

jyn514 commented on src/web/releases.rs:
```diff
@@ -528,13 +564,11 @@ pub fn search_handler(req: &mut Request) -> IronResult<Response> {
                      OFFSET FLOOR(RANDOM() * 280) LIMIT 1",
                     &[]
                 ));
```
This whole thing is pretty clearly incorrect (#789). Probably best to keep functionality changes separate from refactors though.

Kixiron replied:
I didn't change functionality any

jyn514 commented on src/web/releases.rs:
```diff
@@ -528,13 +564,11 @@ pub fn search_handler(req: &mut Request) -> IronResult<Response> {
-                let name: String = rows.get(0).get(0);
-                let version: String = rows.get(0).get(1);
-                let target_name: String = rows.get(0).get(2);
+                let row = rows.into_iter().next().unwrap(); // TODO: Is this really infallible?
```
> Is this really infallible?

As long as there is a crate with >= 100 stars that succeeded to build. So it might fail locally but it will never fail in prod. We can redo the logic as part of fixing #789 but it's fine for now.
